Document Scheduler and Worker state machine

crusaderky opened the pull request:
Closes #5413

Rendered preview:
https://distributed--6948.org.readthedocs.build/en/6948/scheduling-state.html
https://distributed--6948.org.readthedocs.build/en/6948/worker-state.html

crusaderky commented on docs/source/scheduling-state.rst:
```diff
@@ -60,33 +60,41 @@ Tasks flow along the following states with the following allowed transitions:
+    scheduler as well. As soon as a task reaches this state, it is immediately
+    dereferenced from the scheduler.
+
+.. note::
+    There's no intermediate state between waiting ``waiting`` and `no-worker``: as soon
+    as a task has all of its dependencies in memory somewhere on the cluster, it is
+    immediately assigned to a worker. This can lead to very long task queues on the
+    workers, which are then rebalanced dynamically through :doc:`work-stealing`.
```
This will no longer be valid after #6631 
CC @gjoseph92

crusaderky commented on docs/source/scheduling-state.rst:
```diff
@@ -60,33 +60,41 @@ Tasks flow along the following states with the following allowed transitions:
+    scheduler as well. As soon as a task reaches this state, it is immediately
+    dereferenced from the scheduler.
+
+.. note::
+    There's no intermediate state between ``waiting`` / ``no-worker`` and `processing``:
+    as soon as a task has all of its dependencies in memory somewhere on the cluster, it
+    is immediately assigned to a worker. This can lead to very long task queues on the
+    workers, which are then rebalanced dynamically through :doc:`work-stealing`.
```
This will no longer be valid after https://github.com/dask/distributed/issues/6631
CC @gjoseph92

crusaderky commented on docs/source/scheduling-state.rst:
```diff
@@ -60,33 +60,42 @@ Tasks flow along the following states with the following allowed transitions:
+    scheduler as well. As soon as a task reaches this state, it is immediately
+    dereferenced from the scheduler.
+
+.. note::
+    There's no intermediate state between ``waiting`` / ``no-worker`` and
+    ``processing``: as soon as a task has all of its dependencies in memory somewhere on
+    the cluster, it is immediately assigned to a worker. This can lead to very long task
+    queues on the workers, which are then rebalanced dynamically through
+    :doc:`work-stealing`.
```
This will no longer be valid after https://github.com/dask/distributed/issues/6631
CC @gjoseph92

gjoseph92 replied:
Thanks, I started updating this too a little bit over there: 842ee7185912cce663bbfc32fcb624c4c01e9c29

crusaderky replied:
I suggest that we get either PR in first and then update the other one to reflect the new situation

fjetter commented:
@martindurant @jakirkham  you might be interested

gjoseph92 commented on docs/source/worker-state.rst:
```diff
@@ -0,0 +1,433 @@
+-----------------
+The Worker may receive a request to release a key while it is currently in ``flight``,
+``executing``, or ``long-running``. Due to technical limitations around cancelling
+Python threads, such an event cannot cause the related asyncio task to be immediately
```
```suggestion
Python threads, and the way data fetching from peer workers is currently implemented, such an event cannot cause the related asyncio task to be immediately
```
"asyncio task" is a slightly confusing term to use here to me. It's appropriate for the `gather_dep`, but I associate executing tasks with threads, not asyncio. I know there's an `execute` coroutine running, but we _could_ easily `.cancel()` that if we wanted to—the limitation is that it can't cancel its thread.

Maybe plain "task" instead of "asyncio task"? I know that's a bit overloaded with user tasks and `TaskState`s as well. "asyncio task or thread"? "underlying operation"?

I understand what you mean right now since I'm familiar with the current implementation and the existence of `execute`, just wondering if someone unfamiliar with the code would be confused why we were only talking about asyncio.

gjoseph92 commented on docs/source/worker-state.rst:
```diff
@@ -0,0 +1,433 @@
+    and forget that cancellation ever happened;
+  - The scheduler asks the Worker to compute a task that is currently
+    ``cancelled(executing)`` or ``cancelled(long-running)``. The Worker will completely
+    disregard the new :attr:`~TaskState.run_spec` (if it changed) and just switch back
+    to the :attr:`~TaskState.previous` state.
```
```suggestion
    disregard the new :attr:`~TaskState.run_spec` (if it changed) and, switch back
    to the :attr:`~TaskState.previous` state, and wait for the already-executing thread to finish.
```

gjoseph92 commented on docs/source/worker-state.rst:
```diff
@@ -0,0 +1,433 @@
+- The scheduler switches back to its original request:
+
+  - The scheduler asks the Worker to fetch a task that is currently
+    ``cancelled(flight)``; at which point the task will immediately revert to ``flight``
+    and forget that cancellation ever happened;
```
```suggestion
    ``cancelled(flight)``, at which point the task will immediately revert to ``flight``,
    forget that cancellation ever happened, and continue waiting on the data fetch that's
    already running;
```
Slightly redundant, but I think it helps clarify the point

gjoseph92 commented on docs/source/worker-state.rst:
```diff
@@ -0,0 +1,433 @@
+2. Task x is computed successfully on worker A.
+3. When task x transitions to memory on worker A, the scheduler asks worker B to compute
+   task y, which depends on task x.
+4. B starts acquiring the data for x from A, which sends the task into ``flight`` mode.
```
```suggestion
4. B starts acquiring the key x from A, which sends the task into ``flight`` mode.
```
To me, the data would be "for" y, but the data is x.

gjoseph92 commented on docs/source/worker-state.rst:
```diff
@@ -0,0 +1,433 @@
+  and only when the message reaches the worker it will be released there too.
+
+
+Flow control
```
Reading this section of the docs makes me so happy with the design of the worker state machine. Having state transformation strongly separated out from IO and concurrency like this is so nice, and such a big improvement. Nice work!

gjoseph92 reviewed (commented):
Excellent! Very clear and useful documentation; this would be great to read as a new developer interested in working on the Worker. All comments are just naming/grammar nits.

martindurant commented:
This is a great document to have available for reference.

I have a couple of high-level thoughts before getting into detail. None of these mean I am requesting changes in the implementation.

- I do not see the need for the RESUMED state, why not just go to the target state?
- Similarly for CONSTRAINED, which seems identical to READY, which is also constrained but on the thread pool
- Reschedule appears to cause a task to be rescheduled, but forgotten? From https://distributed.dask.org/en/stable/api.html#distributed.Reschedule I understand that the point is to clear its state so that the scheduler can restart its life-cycle.

Some diagrams are disjoint. It makes it confusing to follow. For example, the big diagram at the top of Computing shows rescheduled->released->forgotten, but two diagrams later we see that ERROR and MEMORY have exactly the same paths.

I would change some names to make them clearer, if more verbose is allowed. Something like
WAITING -> WAITING_ON_DEPS_TO_COMPLETE
READY -> READY_TO_RUN
FETCH -> WAITING_TO_FETCH_DEPS
FLIGHT -> DEPS_IN_FLIGHT
RESCHEDULE -> RESCHEDULE_RAISED

I would add a clear and specific definition and consequence of every state. Some of this is in TaskStates, but I would add specific details about the data structures affected.
For example:
- EXECUTING, the associated function is currently running on a thread on this worker and appears as a value of that thread ID in worker.active_threads
- MEMORY, this key is contained in the worker.data dict, the value being the result returned from executing the task's function
- FORGOTTEN, this key is no longer in worker.tasks (or any other structure?) and will soon be garbage collected
- ERROR, an uncaught exception was raised either during the execution of a tasks function, or during serializaation/deserialization. The content of the exception and traceback are held in the task, so that they are relayed to the client and raised there when the client requests the corresponding future's result. Tasks that depend on this task will also get the status ERROR (?).

Why is the initial state of any task apparently RELEASED? 

There is no mention anywhere of Actors, even though a lot of code is dedicated to them.

gjoseph92 commented:
I also just realized I don't think you mention the `TaskState.done` attribute in here. I think it's worth noting since the name is confusing (it sounds like it refers to the task being in a terminal state, as opposed to the execute/fetch coroutine that was responsible for it being complete).

crusaderky commented on docs/source/images/worker-execute-state.dot:
```diff
@@ -3,12 +3,20 @@ digraph{
+    "long-running" -> rescheduled;
+    rescheduled -> released2;
+    released2 -> forgotten;
 }
```
Since the whole cancellation is a special case, I'd rather keep it separate from the general case here to avoid confusing people

crusaderky commented on docs/source/scheduling-state.rst:
```diff
@@ -276,17 +286,52 @@ Transitions occur from stimuli, which are state-changing messages to the
+    Note that this is in addition to the worker's log, which the client can fetch on
+    request (up to a certain length).
+keep-alive
+    A worker informs that it's still online and responsive
```
Added explanation

crusaderky commented on docs/source/worker-state.rst:
```diff
@@ -0,0 +1,433 @@
+    current worker and it's waiting for an available thread.
+constrained
+    Like ``ready``, but the user specified :doc:`resource constraints <resources>` for
+    this task.
```
Not necessarily - resources may be free, but the threads may be not.

crusaderky commented:
> I also just realized I don't think you mention the `TaskState.done` attribute in here. I think it's worth noting since the name is confusing (it sounds like it refers to the task being in a terminal state, as opposed to the execute/fetch coroutine that was responsible for it being complete).

I overhauled the docstring of the flag.

> * I do not see the need for the RESUMED state, why not just go to the target state?

Because you'd have a task in `flight` which is actually taking up a thread from the threadpool, and which will terminate with one of the subclasses of `ExecuteDoneEvent` instead of `GatherDepDoneEvent`; or you'd have a task in `executing` or `long-running` which is actually taking up network resources and will terminate with `GatherDepDoneEvent` instead of `ExecuteDoneEvent`. 

Additionally, when such a task fails, you need to try doing what the scheduler originally asked for. This is actually the norm when a worker dies and the scheduler notices before its peers:
1. worker dies
2. scheduler notices and cancels fetches for all keys which existed exclusively on the dead worker
3. scheduler sends a `compute-task` message to a random worker
4. if the worker receiving the `compute-task` message just happens to be one that previosly had the task in flight, it will need to wait for the TCP connection to fall over, thus sending `GatherDepNetworkFailureEvent` to the state machine, *and then* start the computation.

This use case is also documented in the sphinx documents linked above.

> * Similarly for CONSTRAINED, which seems identical to READY, which is also constrained but on the thread pool

It's a lot more efficient to have two different pipelines for tasks with resources and tasks without, so that a task without resources is not blocked by tasks with.

> * Reschedule appears to cause a task to be rescheduled, but forgotten? From https://distributed.dask.org/en/stable/api.html#distributed.Reschedule I understand that the point is to clear its state so that the scheduler can restart its life-cycle.

reschedule causes the task to be immediately forgotten on the worker and released on the scheduler, which restarts its life-cycle.

> Some diagrams are disjoint. It makes it confusing to follow. For example, the big diagram at the top of Computing shows rescheduled->released->forgotten, but two diagrams later we see that ERROR and MEMORY have exactly the same paths.

Yes, this is on purpose to highlight how reschedule *immediately* transitions to released and forgotten, while error/memory won't transition to released until the scheduler asks to. I updated the diagrams and the "Forgetting tasks" section.

> I would change some names to make them clearer, if more verbose is allowed. 

This would make things seriously hard to read considering how many times these labels appear throughout the code.
It would also make it necessary to use enums to avoid misspellings, further aggravating the code verbosity.

> FETCH -> WAITING_TO_FETCH_DEPS 
> FLIGHT -> DEPS_IN_FLIGHT

No, when a task is in `fetch` or `flight` mode it's itself waiting to be fetched or in flight
If a task is waiting for its dependencies to be gathered, it's in `waiting` state.

> I would add a clear and specific definition and consequence of every state. Some of this is in TaskStates, but I would add specific details about the data structures affected. For example: [...]

Added clarifications.

> Why is the initial state of any task apparently RELEASED?

Historical reasons. There used to be two separate states, `new` at the beginning and `released` at the end; then they were merged into one. Either name is improper for the other state. However, you *can* find a task in `released` state when it's towards the end of its lifetime, whereas a brand new task will immediately transition to `waiting` or `fetch` within a single `transitions()` call.

> There is no mention anywhere of Actors, even though a lot of code is dedicated to them.

Actually, Actors have nothing to do whatsoever with the worker state machine - they're just a task like any other. They are handled exclusively in `Worker`.

crusaderky commented:
All review comments have been addressed

fjetter reviewed (approved):
Good job!
